### PR TITLE
Soft Actor-Critic

### PR DESCRIPTION
**Pull request by kengz**
## Feature / Fix
- implement Soft Actor-Critic (SAC) from https://arxiv.org/abs/1801.01290
- split and update roboschool benchmark specs
- update random baselines for roboschool
- change continuous policy std sampling from `softplus` to log-clamp-exp transform. verified working on PPO BipedalWalker
- add discrete control to SAC using [reparametrizable Categorical distribution (Gumbel-Softmax)](https://pytorch.org/docs/stable/distributions.html#relaxedonehotcategorical)

### Roboschool (continuous control) Benchmark

>Note that the Roboschool reward scales are different from MuJoCo's.

| Env. \ Alg. | A2C (GAE) | A2C (n-step) | PPO | SAC |
|:---|---|---|---|---|
| RoboschoolAnt | | | | 1153.87 <details><summary><i>graph</i></summary></details> |
| RoboschoolHalfCheetah | | | | 1204.68 <details><summary><i>graph</i></summary></details> |
| RoboschoolHopper | | | | 1161.24 <details><summary><i>graph</i></summary></details> |
| RoboschoolWalker2d | | | | 695.36 <details><summary><i>graph</i></summary></details> |

### LunarLander (discrete control) Benchmark

| | |
|---|---|
|sac_lunar_t0_trial_graph_mean_returns_vs_frames|sac_lunar_t0_trial_graph_mean_returns_ma_vs_frames|
| Trial graph | Moving average |

### CODE REVIEW

File: slm_lab/spec/benchmark/sac/sac_lunar.json [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+        "hid_layers": [64, 64, 32],
+        "hid_layers_activation": "relu",
+        "init_fn": "orthogonal_",
+        "clip_grad_val": 0.5,
+        "loss_spec": {
+          "name": "MSELoss"
+        },
+        "optim_spec": {
+          "name": "Adam",
+          "lr": 0.01
```
Comment (CarloLucibello) [inline slm_lab/spec/benchmark/sac/sac_lunar.json]: wrong lr?

**Comment by CarloLucibello** [comment]
Hi, just wanted to point out, the follow-up paper by the authors of SAC
https://arxiv.org/abs/1812.05905
One of the main differences with the original paper is that they don't use a separate V network

File: slm_lab/spec/benchmark/sac/sac_lunar.json (hunk 1 above)
Comment (kengz) [reply slm_lab/spec/benchmark/sac/sac_lunar.json]: 0.005 works better for LunarLander. solved within 200k frames

**Comment by kengz** [comment]
> Hi, just wanted to point out, the follow-up paper by the authors of SAC
> https://arxiv.org/abs/1812.05905
> One of the main differences with the original paper is that they don't use a separate V network

@CarloLucibello Thanks for pointing that out. This PR implements the first version of SAC and adds a discrete control version. I'll implement the improved version in the next PR.
